### PR TITLE
[spmd_types] FLUX enablement

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #3823

### CODE REVIEW

File: torchtitan/models/flux/model/layers.py [hunk 1]
```diff
@@ -133,14 +137,17 @@ class Config(Module.Config):
         norm: QKNorm.Config
         num_heads: int = 8
         qkv_bias: bool = False
+        inner_attention: ScaledDotProductAttention.Config = field(
+            default_factory=ScaledDotProductAttention.Config
+        )
```
Comment (pianpwk) [inline torchtitan/models/flux/model/layers.py:153]: this is so the sharding config can be patched i believe

File: torchtitan/models/flux/parallelize.py [hunk 2]
```diff
@@ -24,13 +30,22 @@
...
 from torchtitan.distributed.fsdp import (
     disable_fsdp_gradient_division,
     enable_fsdp_symm_mem,
 )
+from torchtitan.distributed.full_dtensor import resolve_fsdp_mesh
 from torchtitan.tools.logging import logger
 
 
+def annotate_dp_cp_params_as_r(model: nn.Module, parallel_dims: ParallelDims) -> None:
+    # TODO(pianpwk): Infer these from the active SPMD mesh instead.
```
Comment (pianpwk) [inline torchtitan/models/flux/parallelize.py]: R@DP-axis inference works with partial annotations, but not yet with empty tensors. working to add that in pytorch

File: torchtitan/models/flux/sharding.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        return
+
+    sequence_type = {
+        DP: spmd.S(0),
+        CP: spmd.S(1),
+    }
+    batch_type = {
+        DP: spmd.S(0),
+        CP: spmd.R,
+    }
```
Comment (pianpwk) [inline torchtitan/models/flux/sharding.py:100]: seems flux multimodal inputs are evenly sharded?

File: torchtitan/models/flux/model/layers.py [hunk 4]
```diff
@@ -260,15 +272,15 @@ def forward(
...
-        q = torch.cat((txt_q, img_q), dim=1)
-        k = torch.cat((txt_k, img_k), dim=1)
-        v = torch.cat((txt_v, img_v), dim=1)
+        q = local_concat_text_image_attention_states(txt_q, img_q)
+        k = local_concat_text_image_attention_states(txt_k, img_k)
+        v = local_concat_text_image_attention_states(txt_v, img_v)
 
         q, k = apply_rope(q, k, pe)
-        attn = self.inner_attention(q, k, v)
+        attn = self.inner_attention(q, k, v, is_causal=False)
```
Comment (pianpwk) [inline torchtitan/models/flux/model/layers.py:304]: diffusion model?

Comment (tianyu-l) [reply torchtitan/models/flux/model/layers.py:304]: SelfAttention has this flag=False. I'm not sure why SingleStreamBlock and DoubleStreamBlock don't set it to False. Could you check BFL's official FLUX.1 implementation to confirm?

File: torchtitan/models/flux/trainer.py [hunk 5]
```diff
@@ -58,11 +60,16 @@ def __init__(self, config: Config):
         # Set random seed, and maybe enable deterministic mode
         # (mainly for debugging, expect perf loss).
         # For Flux model, we need distinct seed across FSDP ranks to ensure we randomly dropout prompts info in dataloader
+        distinct_seed_mesh_dims = (
+            ["cp", "dp_shard", "dp_replicate"]
+            if config.parallelism.spmd_backend == "spmd_types"
+            else ["fsdp", "dp_replicate"]
```
Comment (tianyu-l) [inline torchtitan/models/flux/trainer.py:66]: is there anything particular with the order? (it looks inner -> outer mesh ordered, not sure if matters)

File: torchtitan/models/flux/sharding.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/models/flux/sharding.py:100]: please verify that numerics doesn't change with DP & CP involved

File: torchtitan/models/flux/model/layers.py [hunk 6]
```diff
@@ -6,13 +6,17 @@
...
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 
 import torch
 from einops import rearrange
 from torch import nn, Tensor
 from torchtitan.models.common.attention import ScaledDotProductAttention
 from torchtitan.models.common.nn_modules import GELU, LayerNorm, Linear, RMSNorm, SiLU
+from torchtitan.models.flux.model.model import (
+    local_concat_text_image_attention_states,
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/layers.py]: why not defining this function in this file? actually in the same `forward` where this function is used

File: torchtitan/models/flux/model/model.py [hunk 7]
```diff
@@ -6,21 +6,54 @@
...
+    )
+)
+def local_split_text_image(
+    combined: torch.Tensor,
+    text_seq_len: int,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    return combined[:, :text_seq_len], combined[:, text_seq_len:]
+
+
+@spmd.local_map(
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/model.py]: if using local_map, we should annotate `in_types`

However, according to our previous discussion, this could be an ad hoc local context (instead of introducing a function). Either way is fine to me.

File: torchtitan/models/flux/model/model.py [hunk 8]
```diff
@@ -6,21 +6,54 @@
...
+    out_types=spmd.PartitionSpec("dp", "cp", None)
+)
+def local_concat_text_image(text: torch.Tensor, image: torch.Tensor) -> torch.Tensor:
+    return torch.cat((text, image), dim=1)
+
+
+@spmd.local_map(
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/model.py]: similar

File: torchtitan/models/flux/sharding.py (hunk 3 above)
Comment (pianpwk) [reply torchtitan/models/flux/sharding.py:100]: I needed to patch this to do simple KV allgather instead of the pytorch ring-attention context, but yes with that we get bitwise: https://github.com/pytorch/torchtitan/blob/c2a3293fb925c033ccf992b1b11c43f4c7e53684/torchtitan/distributed/context_parallel.py#L80-L99

was going to ask if you wanted that added, or an explicit ring-attention impl in titan

File: torchtitan/models/flux/model/layers.py (hunk 4 above)
Comment (pianpwk) [reply torchtitan/models/flux/model/layers.py:304]: yes, seems both blocks call `attention`: https://github.com/black-forest-labs/flux/blob/802fb4713906133fcbd0d8dc5351620ca4773036/src/flux/modules/layers.py#L194

which calls F.SDPA: https://github.com/black-forest-labs/flux/blob/main/src/flux/math.py#L6

which defaults `is_causal=False`: https://docs.pytorch.org/docs/2.12/generated/torch.nn.functional.scaled_dot_product_attention.html

File: torchtitan/models/flux/sharding.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/models/flux/sharding.py:100]: Let's pretend there's no ring-attention impl.

For the all-gather one, is SDPA supported
- for full bidirectional masking, if q (sharded) and kv (replicate) don't have the same length, would it just work? I think this is doable, but I'm not sure if sdpa supports that.
- for full causal, would sharded q and replicate kv just work? I couldn't imagine how indices are aligned.

Comment (pianpwk) [reply torchtitan/models/flux/sharding.py:100]: > if q (sharded) and kv (replicate) don't have the same length, would it just work?

Based on doc notation it seems supported... https://docs.pytorch.org/docs/2.12/generated/torch.nn.functional.scaled_dot_product_attention.html

> for full causal, would sharded q and replicate kv just work? I couldn't imagine how indices are aligned.

is this concerning FLUX or something else? The callsites in FLUX should all be no-mask right, while other models are on Flex/VarLen I think?

Comment (tianyu-l) [reply torchtitan/models/flux/sharding.py:100]: concerning something that doesn't exist in torchtitan
- if someone uses torchtitan to experiment with causal sdpa with all-gather CP, would they get an error?
- some graph trainer tests are still on SDPA I believe.

Comment (pianpwk) [reply torchtitan/models/flux/sharding.py:100]: cc @fegin , i'm not sure. I guess we could easily introduce a `apply_cp_kv_allgather_to_forward`

Comment (fegin) [reply torchtitan/models/flux/sharding.py:100]: Full causal won't work today without some mask/bias work. This only works for Flux because it doesn't use causal masks.

File: torchtitan/models/flux/trainer.py (hunk 5 above)
Comment (pianpwk) [reply torchtitan/models/flux/trainer.py:66]: I think the seed offset logic here wants innermost-dim first? https://github.com/pytorch/torchtitan/blob/0052a3870c42263b8363972e2de8cbcbdc51def0/torchtitan/distributed/utils.py#L245-L259

not 100% sure

Comment (fegin) [reply torchtitan/models/flux/trainer.py:66]: The order doesn't matter. This is used to guarantee that we will get different seed among these dimenions.

File: torchtitan/models/flux/model/layers.py [hunk 9]
```diff
@@ -16,6 +17,45 @@
...
+
+
+@spmd.local_map(
+    in_types=(
+        spmd.PartitionSpec("dp", "cp", None, None),
+        spmd.PartitionSpec("dp", "cp", None, None),
+    ),
+    out_types=spmd.PartitionSpec("dp", "cp", None, None),
+)
+def local_concat_text_image_attention_states(
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/layers.py]: can we move this within the function where it's used?

File: torchtitan/models/flux/model/layers.py [hunk 10]
```diff
@@ -16,6 +17,45 @@
...
 
 
+@spmd.local_map(
+    in_types=(
+        spmd.PartitionSpec("dp", "cp", None),
+        spmd.PartitionSpec("dp", "cp", None),
+    ),
+    out_types=spmd.PartitionSpec("dp", "cp", None),
+)
+def local_concat_text_image(text: torch.Tensor, image: torch.Tensor) -> torch.Tensor:
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/layers.py]: similar

File: torchtitan/models/flux/trainer.py [hunk 11]
```diff
@@ -58,11 +60,16 @@ def __init__(self, config: Config):
         # Set random seed, and maybe enable deterministic mode
         # (mainly for debugging, expect perf loss).
         # For Flux model, we need distinct seed across FSDP ranks to ensure we randomly dropout prompts info in dataloader
```
Comment (tianyu-l) [inline torchtitan/models/flux/trainer.py:62]: See this comment:
given CP shards data after dataloading (this is TRUE, right??), I don't think we need to put CP in this list, regardless of spmd_types or DTensor.

So technically it should be [dp_shard, dp_replicate]

File: torchtitan/models/flux/parallelize.py [hunk 12]
```diff
@@ -44,21 +61,30 @@ def parallelize_flux(
...
-    dp_mesh = parallel_dims.get_activated_mesh(["dp_replicate", "fsdp"])
-    assert dp_mesh is not None
+    if parallelism.spmd_backend == "spmd_types":
+        dp_mesh, dp_mesh_dims = resolve_fsdp_mesh(parallel_dims)
+    else:
+        dp_mesh = parallel_dims.get_activated_mesh(["dp_replicate", "fsdp"])
+        dp_mesh_dims = None
+        assert dp_mesh is not None
+    if parallelism.spmd_backend == "spmd_types":
+        annotate_dp_cp_params_as_r(model, parallel_dims)
```
Comment (tianyu-l) [inline torchtitan/models/flux/parallelize.py]: wait, these should happen in set_sharding?

Comment (pianpwk) [reply torchtitan/models/flux/parallelize.py]: where is this function? if you mean ShardingConfig-based, it's a bit plain to annotate everything as R@DP/CP

File: torchtitan/models/flux/sharding.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    for block_cfg in config.double_blocks:
+        set_flux_inner_attention_local_map(block_cfg.img_attn.inner_attention)
+        set_flux_inner_attention_local_map(block_cfg.txt_attn.inner_attention)
+        set_flux_inner_attention_local_map(block_cfg.inner_attention)
+
+    for block_cfg in config.single_blocks:
+        set_flux_inner_attention_local_map(block_cfg.inner_attention)
+
+
+def annotate_dp_cp_params_as_r(model: nn.Module, parallel_dims: ParallelDims) -> None:
```
Comment (tianyu-l) [inline torchtitan/models/flux/sharding.py:73]: Did you and Ed finally decided not to infer this automatically inside FSDP?

File: torchtitan/models/flux/model/layers.py [hunk 14]
```diff
@@ -233,13 +254,29 @@ def __init__(self, config: Config):
...
+
+    @staticmethod
+    @spmd.local_map(
+        in_types=(
+            spmd.PartitionSpec("dp", "cp", None, None),
+            spmd.PartitionSpec("dp", "cp", None, None),
+        ),
+        out_types=spmd.PartitionSpec("dp", "cp", None, None),
+    )
+    def local_concat_text_image_attention_states(
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/layers.py]: Does `@spmd.local_map` not support annotating an inline embedded function?

If you have to put it as a separate function

```suggestion
    def _local_concat_text_image_attention_states(
```

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/models/flux/sharding.py (hunk 13 above)
Comment (pianpwk) [reply torchtitan/models/flux/sharding.py:73]: No, we added it, but it only supports the partially-annotated case (TP is on, but not DP/CP), this is the totally unannotated case (param empty, need to look up `spmd.current_mesh`).

Started trying to extend this in pytorch

File: torchtitan/models/flux/model/model.py [hunk 15]
```diff
@@ -142,6 +147,17 @@ def __init__(self, config: Config):
...
 
+    @staticmethod
+    @spmd.local_map(
+        in_types=(
+            spmd.PartitionSpec("dp", "cp", None),
+            spmd.PartitionSpec("dp", "cp", None),
+        ),
+        out_types=spmd.PartitionSpec("dp", "cp", None),
+    )
+    def local_concat_text_image(text: Tensor, image: Tensor) -> Tensor:
```
Comment (tianyu-l) [inline torchtitan/models/flux/model/model.py]:
```suggestion
    def _local_concat_text_image(text: Tensor, image: Tensor) -> Tensor:
```

Comment (tianyu-l) [reply torchtitan/models/flux/model/model.py]: btw could you answer my question?

> Does `@spmd.local_map` not support annotating an inline embedded function?

Comment (pianpwk) [reply torchtitan/models/flux/model/model.py]: oops sorry, dropped a change. should be fine
